pkg/csi/service/node.go

```diff
@@ -674,109 +683,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
-		if err != nil {
+func fetchTopologyLabelsUsingCR(ctx context.Context, nodeName, nodeID string) (map[string]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Get kube config.
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal, "failed to fetch kubeconfig. Error: %+v", err)
+	}
+	// Create controller runtime client.
+	controllerRuntimeClient, err := k8s.NewClientForGroup(ctx, config, csinodetopologyv1alpha1.GroupName)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create controllerRuntimeClient. Error: %+v", err)
+	}
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, nodeName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind: "Node",
+					Name: nodeObj.Name,
+					UID: nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: nodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = controllerRuntimeClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
 			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to get VirtualCenterConfig from cns config. err: %v", err)
+				"failed to create CSINodeTopology CR. Error: %+v", err)
 		}
-		vcManager := cnsvsphere.GetVirtualCenterManager(ctx)
-		vcenter, err := vcManager.RegisterVirtualCenter(ctx, vcenterconfig)
-		if err != nil {
-			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to register vcenter with virtualCenterManager. err: %v", err)
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", nodeName)
+	}
+
+	// Watch on the CSINodeTopology CR.
+	csiNodeTopologyWatcher, err := k8s.NewCSINodeTopologyWatcher(ctx, config)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create a watcher for CSINodeTopology CR. Error: %+v", err)
+	}
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
+	watchCSINodeTopology, err := csiNodeTopologyWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.SelectorFromSet(fields.Set{"metadata.name": nodeID}).String(),
+		TimeoutSeconds: &timeoutSeconds,
+		Watch:          true,
+	})
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to watch on CSINodeTopology instance with name %q. Error: %+v", nodeName, err)
+	}
+	defer watchCSINodeTopology.Stop()
+
+	// Check if status gets updated in the instance within the given timeout seconds.
+	for event := range watchCSINodeTopology.ResultChan() {
```

Comment:
Is it possible for the instance to already exist? In this case, does this channel get an event?

Reply:
Yes it will still get an event. I guess the behavior of this watch is similar to the output of the command `kubectl get csinodetopology <nodeName> -w` which gives the current state of the instance and then keeps updating it as & when there are changes made to it.